Performance optimization: Move all LittleEndianDataInputStream functionality into ByteBufferInputStream

theosib-amazon opened the pull request:
I broke up https://github.com/apache/parquet-mr/pull/953 into more digestible pieces. This new PR is the lowest level set of changes. By themselves, these additions to ByteBufferInputStream don't yield much improvement, so future PRs will include modifications to other source files that take advantage of this new functionality.

The complete set of changes (including subsequent PRs) is for performance optimization. In benchmarking with Trino, we find query performance to improve from 5% to 15%, depending on the query, and that includes all the I/O time from S3.

All of LittleEndianDataInputStream functionality is moved into ByteBufferInputStream, without changing any pre-existing interfaces or functionality. These changes yield the following benefits:
- Elimination of extra layers of abstraction and method call overhead
- Enable the use of intrinsics for readInt, readLong, etc.
- Availability of faster access methods like readFully and skipFully, without the need for helper functions

This PR also marks LittleEndianDataInputStream as deprecated.

Context:
I've been working on improving Parquet reading performance in Trino, mostly by profiling while running performance benchmarks and TPCDS queries. This PR is a subset of the changes I made that have more than doubled the performance of a lot of TPCDS queries (wall clock time, including the S3 access time). If you are kind enough to accept these changes, I look forward to offering further performance improvements.

theosib-amazon committed:
Move all LittleEndianDataInputStream functionality into ByteBufferInputStream

To improve performance, all multi-byte access functionality from
LittleEndianDataInputStream has been merged into ByteBufferInputStream.
LittleEndianDataInputStream is marked deprecated.

theosib-amazon committed:
Unhacked thrift version requirement in pom

theosib-amazon committed:
Update MultiBufferInputStream.java

Made exception catch more specific for read() and readByte().

theosib-amazon committed:
Added tests for ByteBufferInputStream

Also fixed bug discovered by that testing

theosib-amazon commented:
I added some tests for this new code.

rdblue commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -28,6 +28,16 @@
 import org.apache.parquet.ShouldNeverHappenException;
 
+/*
+Changes implemented:
```
Can you remove this? Generally, we want PRs that can be merged. You can add comments about what is changing in the PR rather than in the code.

theosib-amazon replied:
Removed

rdblue commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -49,6 +59,19 @@ public static ByteBufferInputStream wrap(List<ByteBuffer> buffers) {
     }
   }
 
+  public static ByteBufferInputStream wrap(ByteBuffer buffer, int offset, int count) {
```
What is the purpose of these methods? We already wrap buffers, why not have the caller handle this?

It's also really strange to pass a buffer with an offset and a "count" (do you mean length?) because buffers already have that information built in.

theosib-amazon replied:
This was to make ByteBufferInputStream resemble ByteBuffer. ByteBuffer has methods like this. I've gone ahead and removed them at your request.

rdblue commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -74,11 +97,26 @@ public ByteBufferInputStream(ByteBuffer buffer) {
     byteBuf.limit(count);
     delegate = wrap(byteBuf);
+    // ... but it would probably be faster to do this:
+//    delegate = wrap(buffer, offset, count);
```
Please remove notes to self.

theosib-amazon replied:
Done.

theosib-amazon committed:
Update ByteBufferInputStream.java

Removed unnecessary comment
Removed unused wrapper method

theosib-amazon committed:
Update ByteBufferInputStream.java

Removed unused wrappers and constructors

theosib-amazon committed:
Update ByteBufferInputStream.java

Removed line of code that was commented out instead of properly deleted

theosib-amazon committed:
Update SingleBufferInputStream.java

Cleaned up some comments

rdblue commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -274,6 +329,7 @@ public void mark(int readlimit) {
     if (mark >= 0) {
       discardMark();
     }
+    
```
Nit: unnecessary whitespace change.

rdblue commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -157,4 +181,83 @@ public void reset() throws IOException {
     return delegate.markSupported();
   }
+
+  public void close() throws IOException {
```
Close delegate?

theosib-amazon committed:
Update ByteBufferInputStream.java

Removed some more unused methods

theosib-amazon committed:
Update LittleEndianDataInputStream.java

@Deprecated postpones to future PR

theosib-amazon committed:
Update MultiBufferInputStream.java

Removed unnecessary methods. Reverted whitespace change.

theosib-amazon committed:
Update SingleBufferInputStream.java

Removed unnecessary methods

theosib-amazon committed:
Update MultiBufferInputStream.java

Removed whitespace change

shangxinli commented:
Is the '5% to 15%' gain from this change or along with other changes? If it is later, can you share the point to other changes? Like to see the overall changes before committing.

theosib-amazon commented:
That improvement comes from a larget set of changes. I have a design doc that goes over all those changes plus some more that make it possible to get even more performance improvements.

https://docs.google.com/document/d/1fBGpF_LgtfaeHnPD5CFEIpA2Ga_lTITmFdFIcO9Af-g/edit?usp=sharing

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -157,4 +165,80 @@ public void reset() throws IOException {
+    }
+  }
+
+  public int readUnsignedVarInt() throws IOException {
```
Is it copied from ByteUnites? I wonder why we don't use that directly?>

theosib-amazon replied:
Not exactly. The one in BytesUtils calls methods that read one byte at a time. This one can take advantage of faster methods that read whole words at a time. This is a critical-path method, so it's a performance win to eliminate the extra level of abstraction and all the extra overhead fetching individual bytes and shifting.

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -157,4 +165,80 @@ public void reset() throws IOException {
+    return ((ch3 << 16) + (ch2 << 8) + (ch1 << 0));
+  }
+
+  public int readIntLittleEndianPaddedOnBitWidth(int bitWidth)
```
Is it copied from ByteUnites? I wonder why we don't use that directly?

shangxinli replied:
I meant the method here: https://github.com/apache/parquet-mr/blob/master/parquet-common/src/main/java/org/apache/parquet/bytes/BytesUtils.java#L120. It is exactly same but just different signature.

theosib-amazon replied:
I know about that method. The BytesUtils code always reads one byte at a time. My version will read a whole word at a time for short and int. This is faster.

theosib-amazon replied:
Wait. Are you referring to readIntLittleEndianPaddedOnBitWidth or readIntLittleEndianOnThreeBytes?

The former is definitely faster. An argument could be made to remove the latter, although it'll take longer for the JIT to hide the extra layers of virtual calls.

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -157,4 +165,80 @@ public void reset() throws IOException {
+    return Double.longBitsToDouble(readLong());
+  }
+
+  public int readIntLittleEndianOnThreeBytes() throws IOException {
```
Is it copied from BytesUtils.java? I wonder why we don't use that directly?

theosib-amazon replied:
See my other comments.

shangxinli replied:
Check this https://github.com/apache/parquet-mr/blob/master/parquet-common/src/main/java/org/apache/parquet/bytes/BytesUtils.java#L110

theosib-amazon replied:
See my other comment on this. These two methods have the the same outcome, but mine is faster. I believe this is warranted for a performance critical path.

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -238,8 +257,31 @@ public int read(byte[] bytes, int off, int len) {
   @Override
-  public int read(byte[] bytes) {
-    return read(bytes, 0, bytes.length);
+  public void readFully(byte[] bytes, int off, int len) throws IOException {
```
Can you cast a light why we need to add the implementation readFully() here? For performance improvement?

theosib-amazon replied:
There are situations where we need to read an exact number of bytes and throw an exception if not enough are available. This is faster than reading maybe enough and then checking, and this is a performance critical path.

sunchao replied:
Why can't we just track the remaining bytes of the stream on the client side and check before reading any bytes?

theosib-amazon replied:
Normally, the user of the class would read exactly the right number of bytes. These checks and exceptions exist only to catch bugs elsewhere. This is one reason why it's important to minimize the overhead of these checks in such performance-critical methods.

shangxinli replied:
The difference between this method and read() is mainly to precheck if there is enough remaining length. I believe this can be done by wrapping up the read() method and adding the prechecks. Duplicating the code makes it harder to maintain.

theosib-amazon replied:
I'll make these changes if you insist. But those prechecks are expensive, which is why I'm trying to avoid them when possible in a performance critical path.

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -238,8 +257,31 @@ public int read(byte[] bytes, int off, int len) {
+    }
+
+    int bytesRead = 0;
+    while (bytesRead < len) {
```
This seems duplicate with above line 244.

theosib-amazon replied:
There are two key differences that make it hard to combine them without hurting performance for one, the other, or both, and they're both performance critical.

shangxinli replied:
Duplicating code is hard to maintain.

theosib-amazon replied:
Well, my objective here is to maximize performance. So we have to decide between maintainability and performance. Let's deliberate over this a bit more, and I'll do what you think is best.

wgtmac replied:
> Well, my objective here is to maximize performance. So we have to decide between maintainability and performance. Let's deliberate over this a bit more, and I'll do what you think is best.

I deeply understand this is a difficult trade-off. Do you have any evidence on the performance penalty if we wrap `read` and `readFully` methods to share some common logic? If the penalty is acceptable, we should definitely go for maintainability.

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -379,4 +427,120 @@ public void remove() {
+    
+    if (current.remaining() >= 8) {
+      // If the whole short can be read from the current buffer, use intrinsics
+      this.position += 8;
```
USE Long.BYTES instead of 8

shangxinli commented:
@sunchao Can you have a review?

theosib-amazon committed:
Update MultiBufferInputStream.java

Use constants for byte size of words

sunchao commented:
Is this mostly a refactoring PR? I also don't see `LittleEndianDataInputStream` being marked as deprecated.

theosib-amazon commented:
> Is this mostly a refactoring PR? I also don't see `LittleEndianDataInputStream` being marked as deprecated.

I initially marked `LittleEndianDataInputStream` as deprecated. But I seem to recall that I was advised by Ryan Blue to do that in a later PR.

sunchao commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -89,6 +91,15 @@ public long skip(long n) {
+  @Override
+  public void skipFully(long n) throws IOException {
+    if (current == null || n > length) {
+      throw new EOFException();
```
ditto: some info in exception, better retain the original message: `Not enough bytes to skip: " + skipped + " < " + n)`

theosib-amazon replied:
Got it

sunchao commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -379,4 +427,120 @@ public void remove() {
+  }
+
+  @Override
+  public int readUnsignedByte() throws IOException {
```
this looks the same as `read`

theosib-amazon replied:
You're right. They're identical. I think the C1 compiler will eliminate the overhead of read() calling readUnsignedByte().

shangxinli replied:
If they are identical, we should remove the duplicate one.

theosib-amazon replied:
Looks like I took care of this already.

sunchao commented on parquet-common/src/main/java/org/apache/parquet/bytes/MultiBufferInputStream.java:
```diff
@@ -379,4 +427,120 @@ public void remove() {
+
+  @Override
+  public byte readByte() throws IOException {
+    return (byte)readUnsignedByte();
```
nit: can we add a space after the cast `(byte) readUnsignedByte()`? for all the other places too

theosib-amazon replied:
got it

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -88,6 +136,21 @@ public long skip(long n) {
+    
+    try {
+      buffer.position(buffer.position() + (int)n);
+    } catch (IllegalArgumentException e) {
```
I did it this way on purpose. This way is always faster. Doing the check has to wait until there's profiling info and the C2 compiler gets hold of this code.

theosib-amazon replied:
I already don't like the fact that I have to check the argument to make sure it's not negative and not bigger than int max. buffer.positiion() already checks for the position going out of bounds and throws an exception, so it would be redundant to have another check for the exact same thing here. A catch for an exception that never happens is basically always free, while a test for a condition that never happens is not free until profiling gets enough info about it for the C2 compiler to eliminate it.

shangxinli replied:
the try/catch is also more expensive than checking. I agree with Chao to have the check instead of try/catch.

theosib-amazon replied:
A try/catch is basically free when no exception is thrown, while a check is not. I have tested this, and the try/catch is empirically faster, since the no-exception case is the common case. Putting in the check means we have to wait until the C2 compiler produces a trace without the branch. But even then, there's always the overhead of a check somewhere to be able to fall back to interpretation if the condition is not correct for the trace. I'm avoiding all of that entirely, making this faster in of the most common case.

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -38,6 +39,34 @@ class SingleBufferInputStream extends ByteBufferInputStream {
+    this.buffer.order(java.nio.ByteOrder.LITTLE_ENDIAN);
+  }
+
+  SingleBufferInputStream(byte[] inBuf) {
```
Yeah. I put this here I think because I'm trying to make BufferInputStream have the same functionality as multiple other things that I'm combining into one. Some tests would be good. I'll see about writing some tests soon.

theosib-amazon replied:
I went ahead and added two tests to cover the unused SingleBufferInputStream constructors. I considered just deleting these constructors, but I decided that it might be valuable to include them as documentation on how to do this in a way that is congruent to the behavior of HeapByteBuffer, just in case anyone wanted to do this in the future. There's also the risk that someone would think they have to wrap an array with ByteBuffer before using SingleBufferInputStream, but it would be better to avoid the overhead of ByteBuffer.duplicate(). (ByteBuffer.duplicate() appears to take constant time by making a reference to the same backing array, but it's a big constant with loads of checks.)

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -38,6 +39,34 @@ class SingleBufferInputStream extends ByteBufferInputStream {
+    this.startPosition = 0;
+  }
+
+  SingleBufferInputStream(List<ByteBuffer> inBufs) {
```
IIRC, I used to have a similar constructor for ByteBufferInputStream, but I was advised to remove it, so I had this here for uniformity. We can remove this.

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -70,9 +105,22 @@ public int read(byte[] bytes, int offset, int length) throws IOException {
+  public void readFully(byte[] bytes, int offset, int length) throws IOException {
+    try {
+      buffer.get(bytes, offset, length);
+    } catch (BufferUnderflowException|IndexOutOfBoundsException e) {
```
Got it.

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -88,6 +136,21 @@ public long skip(long n) {
+  @Override
+  public void skipFully(long n) throws IOException {
+    if (n < 0 || n > Integer.MAX_VALUE) {
+      throw new IllegalArgumentException();
```
got it

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -70,9 +105,22 @@ public int read(byte[] bytes, int offset, int length) throws IOException {
   @Override
   public long skip(long n) {
+    if (n < 0) {
+      throw new IllegalArgumentException();
```
got it

theosib-amazon commented on parquet-common/src/main/java/org/apache/parquet/bytes/SingleBufferInputStream.java:
```diff
@@ -174,4 +254,64 @@ public boolean markSupported() {
+  }
+
+  /*
+  Use ByteBuffer.getLonmg(), which takes advantage of platform intrinsics
```
got it

theosib-amazon committed:
Update SingleBufferInputStream.java

Minor code changes on request. Removed redundant code. Fixed code formatting nits. More informative exceptions. Removed constructor that just throws exception.

theosib-amazon committed:
Update MultiBufferInputStream.java

Minor code changes on request. Removed redundant code. Fixed code formatting nits. More informative exceptions.

theosib-amazon committed:
Added tests for two as-yet unused constructors for SingleBufferInputStream.

Rather than deleting these, I decided to keep the constructs as documentation
on how to do these things according to how HeapByteBuffer wraps arrays. And
it would be best if we didn't have to suffer the overhead of creating a
ByteBuffer first in order to create a SingleBufferInputStream.

theosib-amazon committed:
Fixed compile and test failures

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -138,6 +134,18 @@ public int read(byte[] b, int off, int len) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  public void readFully(byte b[]) throws IOException {
```
Why we need to be different between line 137 ' byte[] b' and 141 ' byte b[]'?

theosib-amazon replied:
This is almost certainly the result of a copy/paste of the method signature, and I just didn't notice. I went ahead and fixed it.

shangxinli commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -138,6 +134,18 @@ public int read(byte[] b, int off, int len) throws IOException {
+    readFully(b, 0, b.length);
+  }
+
+  public void readFully(byte b[], int off, int len) throws IOException {
```
Don't see where it is used. Don't know why it is 'public'.

theosib-amazon replied:
This is what the method signatures for DataInputStream.readFully look like.

I also have a whole bunch of other performance improvements I want to contribute (https://docs.google.com/document/d/1fBGpF_LgtfaeHnPD5CFEIpA2Ga_lTITmFdFIcO9Af-g/edit?usp=sharing), and I think this might get used in some of that code.

I'm very soon going to publish an open preview of all of my proposed changes to a branch of my own fork, so we'll be able to check this out.

theosib-amazon replied:
Here's the complete preview of my changes to ParquetMR: https://github.com/theosib-amazon/parquet-mr/tree/batch-read-optimizations

theosib-amazon committed:
Update ByteBufferInputStream.java

Changed byte b[] to byte[] b

wgtmac commented on parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java:
```diff
@@ -157,4 +165,80 @@ public void reset() throws IOException {
+    return delegate.readUnsignedByte();
+  }
+
+  public short readShort() throws IOException {
```
I like the idea to provide these read functions to enable larger read. BTW, is there any use case to read a batch of shorts (and other numeric types)?

theosib-amazon replied:
I news the new batch read methods heavily in some optimizations I made to Trino. As for short, I can't say I recall any uses in Trino of readShorts(). readShort() is used indirectly through a method that reads a variable sized representation.

wgtmac reviewed (commented):
IMHO, the open source community priorities code maintainability more than performance gain as the project is widely adopted and maintained by many users and developers. However, if there is any concrete benchmark result to provide compelling evidence, it will be helpful to the discussion. @theosib-amazon @shangxinli @sunchao

shangxinli commented:
@theosib-amazon Thanks again for your contribution! I see the comments are generally around duplicating code, refactoring, and making code maintainable. If you have a measurement of improvement on this change alone, it would help the reviewers
